src/storage/src/hummock/sstable/delete_range_aggregator.rs

```diff
@@ -0,0 +1,328 @@
+
+use risingwave_hummock_sdk::key::{get_epoch, key_with_epoch, user_key};
+use risingwave_hummock_sdk::key_range::KeyRange;
+use risingwave_hummock_sdk::HummockEpoch;
+
+#[derive(Clone)]
+pub struct DeleteRangeTombstone {
+    start_user_key: Vec<u8>,
+    end_user_key: Vec<u8>,
+    sequence: HummockEpoch,
+}
+
+impl PartialEq<Self> for DeleteRangeTombstone {
+    fn eq(&self, other: &Self) -> bool {
+        self.end_user_key.eq(&other.end_user_key) && self.sequence == other.sequence
+    }
+}
+
+impl PartialOrd for DeleteRangeTombstone {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        let ret = other
+            .end_user_key
+            .cmp(&self.end_user_key)
+            .then_with(|| other.sequence.cmp(&self.sequence));
+        Some(ret)
+    }
+}
+
+impl Eq for DeleteRangeTombstone {}
+
+impl Ord for DeleteRangeTombstone {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.end_user_key
+            .cmp(&other.end_user_key)
+            .then_with(|| other.sequence.cmp(&self.sequence))
+    }
+}
+
+pub struct DeleteRangeAggregator {
+    delete_tombstones: Vec<DeleteRangeTombstone>,
+    key_range: KeyRange,
+    watermark: u64,
+    gc_delete_keys: bool,
+}
+
+impl DeleteRangeAggregator {
+    pub fn new(key_range: KeyRange, watermark: u64, gc_delete_keys: bool) -> Self {
+        Self {
+            key_range,
+            delete_tombstones: vec![],
+            watermark,
+            gc_delete_keys,
+        }
+    }
+
+    pub fn add_tombstone(&mut self, data: Vec<(Vec<u8>, Vec<u8>)>) {
+        for (key, mut end_user_key) in data {
+            let mut start_user_key = user_key(&key);
+            let sequence = get_epoch(&key);
+            if !self.key_range.left.is_empty() {
+                let split_start_user_key = user_key(&self.key_range.left);
+                if split_start_user_key.gt(end_user_key.as_slice()) {
+                    continue;
+                }
+                if split_start_user_key.gt(start_user_key) {
+                    start_user_key = split_start_user_key;
+                }
+            }
+            if !self.key_range.right.is_empty() {
+                let split_end_user_key = user_key(&self.key_range.right);
+                if split_end_user_key.le(start_user_key) {
+                    continue;
+                }
+                let split_end_user_key = user_key(&self.key_range.right);
+                if split_end_user_key.le(start_user_key) {
+                    continue;
+                }
+                if split_end_user_key.lt(end_user_key.as_slice()) {
+                    end_user_key = split_end_user_key.to_vec();
+                }
+            }
+
+            let tombstone = DeleteRangeTombstone {
+                start_user_key: start_user_key.to_vec(),
+                end_user_key,
+                sequence,
+            };
+            self.delete_tombstones.push(tombstone);
+        }
+    }
+
+    pub fn sort(&mut self) {
```

Comment:
Since it is not different from sort condition of tombstone, shall we name it `sort_by_start`?